common/page.go

```diff
@@ -364,6 +363,127 @@ func (p *Page) initEvents() {
 	}()
 }
 
+func (p *Page) urlGroupingName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+	p.eventHandlersMu.RUnlock()
+
+	var name string
+	var nameChanged bool
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p.eventHandlersMu.RLock()
+
+	for _, h := range p.eventHandlers[EventPageMetricCalled] {
+		// A handler can register another handler from within itself. This is
+		// the reason to unlock the mutex before calling the handler.
+		p.eventHandlersMu.RUnlock()
+
+		nameCh := make(chan string)
+		em := &ExportedMetric{
+			ctx:    ctx,
+			urlTag: urlTag,
+			nameCh: nameCh,
+		}
+
+		// handler runs on another goroutine.
+		h(em)
+
+		// We want to wait for the name to come in from the handler. When the
+		// nameCh is closed by the handler the name is returned back upstream.
+		ok := true
+		for ok {
+			select {
+			case n, o := <-nameCh:
+				ok = o
+				if ok {
+					name = n
+					nameChanged = true
```

Comment:
What does `nameChanged` do in this context?

Reply:
Keeps track of the whether the name was changed, and returns true if it has or false. Makes it simpler upstream to then work with the name or not based on whether it is true or false.

Reply:
I guess `name` here means a "tag name." And, `nameChanged` means that the current URL matches to a user-provided URL pattern. So, we change the tag name. By upstream, you mean the metric system. Right? How can we make this flow/logic clearer in the code? It might be obvious to you since you have worked on it—but not others like me 😅

Reply:
Where's the confusion? Is it the names of the variables?

> By upstream, you mean the metric system. Right?
 👍

Reply:
I've updated the variable names: 760a38042cc4c3554cc53f4c002e8343614982f4

Reply:
The confusion arises because of the naming and also the nature of the change. It has many indirections. So, it's not easy to wrap my head around. Naming adds more confusion to the party 😄